### PR TITLE
[av] Fix precision of the player position

### PR DESCRIPTION
**Pull request by jpudysz**
# Why

Hello, Expo Team!

We are developing a complex video editor for our client, utilizing your `EXAVPlayer`. This editor enables users to seek any millisecond of the video. However, we encountered some issues while attempting to use `setPositionAsync`.

# How

As per Apple's documentation (https://developer.apple.com/documentation/avfoundation/avplayer/1387741-seektotime?language=objc), we understand that `toleranceBefore` and `toleranceAfter` can be set. Despite this, we've noticed edge cases where EXAVPlayer fails to navigate to the specified millisecond.

# Test Plan

The issue was tested in a local project. Adjusting the precision resolved the problem related to rounding.

Before:
1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```tsx
videoRef.current?.setPositionAsync(267, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3.  Observe `newPosition` in Objective-C.

image

4. Notice the video did not seek to 267 ms

After:

1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```tsx
videoRef.current?.setPositionAsync(267, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3.  Observe `newPosition` in Objective-C.

image

4. The video now seeks accurately to 267 ms.

# Checklist

No need to update the docs

### CODE REVIEW

**Review by tsapeta — APPROVED** [review]
Good catch! I'll make sure to include it in SDK 50.
Congrats on your first contribution 👏 😉

**Comment by jpudysz** [comment]
Thank you, @tsapeta! I'm looking forward to giving you a high five at the next App.js Conference 😁

**Review by tomekzaw — COMMENTED** [review]
Good catch!

**Comment by jpudysz** [comment]
Hey, I've added another case and updated the PR – please let me know your thoughts on it and if I should open new PR.

While the previous fix resolves my issue for a single video, I've noticed that `EXAVPlayer` sometimes disrupts the synchronization when handling two videos. This issue arises because one line of the code compares a JavaScript number to a `long` value. While rounding here might seem convenient, it occasionally leads to synchronization problems.

```objc
// compare number (int) to long
mustSeek = currentPositionMillis.longValue != [self _getRoundedMillisFromCMTime:_currentPosition].longValue;
```

The proposed solution is to always seek if the user sets the tolerance to zero. This implies that the user acknowledges the potential performance drawbacks. With this adjustment, I can consistently synchronize my two videos down to the exact millisecond.

Again tested locally CC @tsapeta

**Review by lukmccall — APPROVED** [review]

**Comment by jpudysz** [comment]
One more case guys! 😄 

# Why

With the two changes mentioned above, everything works perfectly, and `self.player.currentTime` is always set to the desired millisecond.

Unfortunately, sending the event back to JavaScript involves a different rounding algorithm. Even though the player successfully seeks to the desired millisecond, the reported time is slightly different.

Problem lives in the `_getRoundedMillisFromCMTime` where the value is converted from `CMTime` to `NSNumber`.

```objc
- (NSNumber *)_getRoundedMillisFromCMTime:(CMTime)time
{
  return CMTIME_IS_INVALID(time) || CMTIME_IS_INDEFINITE(time) ? nil : @((long)(CMTimeGetSeconds(time) * 1000));
}
```
My latest change introduces a better (more mathematically accurate) rounding method than the original solution.

# Test Plan

Before:

1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```js
videoRef.current?.setPositionAsync(1001, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3. Observe `positionMillis` inside the `getStatus` method, where we compute the value to be sent to JavaScript.
4. Value is 1000, even though player keeps 1001!



After:

1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```js
videoRef.current?.setPositionAsync(1001, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3. Observe `positionMillis` inside the `getStatus` method, where we compute the value to be sent to JavaScript.
4. Value is 1001

precise

File: packages/expo-av/ios/EXAV/EXAVPlayerData.m [hunk 1]
```diff
@@ -205,6 +205,27 @@ - (BOOL)_shouldPlayerPlay
...
 
+- (BOOL)_hasZeroTolerance:(NSDictionary *)parameters
+{
+  if ([parameters objectForKey:EXAVPlayerDataStatusSeekMillisToleranceBeforeKeyPath] == nil) {
+    return NO;
+  }
+    
+  NSNumber *seekMillisToleranceBefore = parameters[EXAVPlayerDataStatusSeekMillisToleranceBeforeKeyPath];
+    
+  if (CMTimeCompare(CMTimeMakeWithSeconds(seekMillisToleranceBefore.floatValue / 1000, NSEC_PER_SEC), kCMTimeZero) != 0) {
```
[the remark this answers is not in the document]
Comment (jpudysz) [reply packages/expo-av/ios/EXAV/EXAVPlayerData.m:216]: No, these values are fine. There is no need for precise rounding of tolerances.

[the remark this answers is not in the document]
Comment (tsapeta) [reply packages/expo-av/ios/EXAV/EXAVPlayerData.m:216]: I'm also asking about the consistency. I think it makes sense to generally avoid `floatValue` as we already know that it's causing precision issues.

[the remark this answers is not in the document]
Comment (jpudysz) [reply packages/expo-av/ios/EXAV/EXAVPlayerData.m:216]: Surely it won't hurt, if you want I can update those values too

[the remark this answers is not in the document]
Comment (tsapeta) [reply packages/expo-av/ios/EXAV/EXAVPlayerData.m:216]: Yeah, let's update it then 😉

**Review by tsapeta — APPROVED** [review]
I'm sorry, I totally forgot about this PR 😅 Is this ready to merge?

**Comment by jpudysz** [comment]
Yes, it’s ready and working great on my side. There have been no reports from QA for the past 3 weeks 🤞
